resources/type-system/inference.md

```diff
@@ -970,167 +970,232 @@ with respect to `L` under constraints `C0`
+non-normative, so they are typeset in italics._
 
-  - Otherwise, if `T_1` is `dynamic`, then let `m_2` be an expression artifact
-    with static type `T_2`, whose runtime behavior is as follows:
+## Coercions
 
-    - Execute expression artifact `m_1`, and let `o_1` be the resulting value.
+Before considering the specific rules for type inferring each type of
+expression, it is useful to define an operation known as _coercion_. _Coercion_
+is a type inference step that is applied to an elaborated expression `m_1` and a
+target type `T`, and produces a new elaborated expression `m_2`.
 
-    - If `o_1` is an instance of the type `T_2`, `m_2` completes with the value
-      `o_1`. _In other words, the implicit cast from `dynamic` to `T_2` has
-      succeeded. Expression soundness follows trivially because the completed
-      value is an instance of `T_2`, and the static type is `T_2`._
+_The coercion operation satisfies the soundness invariant that the static type
+of `m_2` is guaranteed to be a subtype of `T`. A proof of this is sketched
+below._
+
+_Coercions are used in most situations where the existing spec calls for an
+assignability check._
 
-    - Otherwise, `m_2` completes with an exception. _In other words, the
-      implicit cast from `dynamic` to `T_2` has failed. This trivially satisfies
-      expression soundess, since expression soundness is only concerned with the
-      value of the expression when it completes normally._
+Coercion of an elaborated expression `m_1` to type `T` produces `m_2`, which is
+determined as follows:
 
-  - Otherwise, if `T_1` is an interface type that contains a method called
-    `call` with type `U`, and `U <: T_2`, then let `m_2` be an expression
-    artifact with static type `T_2`, whose runtime behavior is as follows:
+- Let `T_1` be the static type of `m_1`.
 
-    - Execute expression artifact `m_1`, and let `o_1` be the resulting
-      value. _By expression soundness, `o_1` will be an instance of the type
-      `T_1`._
+- If `T_1 <: T`, then let `m_2` be `m_1`. _Since `m_1` and `m_2` are the same,
+  the static type of `m_2` is `T_1`. Therefore, since `T_1 <: T`, the soundness
+  invariant is satisfied._
 
-    - Let `o_2` be the result of tearing off the `call` method of `o_1`. _This
-      is guaranteed to succeed since `o_1` is an instance of `T_1`, and it is
-      guaranteed by tear-off soundness that `o_2` will be an instance of `U`._
+- Otherwise, if `T_1` is `dynamic`, then let `m_2` be
+  `@IMPLICIT_CAST<T>(m_1)`. _Since `@IMPLICIT_CAST<T>(m_1)` has a static type of
+  `T`, the soundness invariant is satisfied._
 
-    - `m_2` completes with the value `o_2`. _Expression soundness follows from
-      the fact that `o_2` is an instance of `U` and `U <: T_2`._
+- Otherwise, if `T_1` is an interface type that contains a method called `call`
+  with type `U`, and `U <: T`, then let `m_2` be `m_1.call`. _Since `m_1.call`
```

Comment:
Interesting. I’m not sure how I feel about this.

On the one hand, using ordinary Dart syntax for elaborated expressions is nice because it allows readers familiar with Dart (who, let’s be honest, constitute my entire target audience) to understand what’s going on more quickly, and allows me to focus my specification efforts on the details of type inference rather than trying to rigorously define the semantics of everything (which I think was one of the drawbacks of my previous approach).

On the other hand, using new syntax for elaborated expressions has the advantage that it’s much more obvious to the reader when we are talking about an elaborated expression vs. an unelaborated expression.

As to your comment that "Since `m_1.call` has static type ..." is undefined, my mental model in this PR is that every elaborated expression has a well-defined static type; when the elaborated expression uses standard Dart syntax, the static type of the expression is defined using the static typing rules in `dartLangSpec.tex`; when it uses one of the new `@` constructs I’ve introduced, the static type of the expression is defined in the section that defines all of those constructs ("New operations allowed in elaborated expressions"). To put it another way, since elaborated expressions are based on standard Dart, with the addition of some new constructs and the prohibition of some old ones, they inherit the static typing rules defined in `dartLangSpec.tex`. Does that seem like a reasonable approach to you? If so, is there something I could do that would help clarify that’s the approach I’m taking?

Note that this is a change from my previous mental model (prior to this CL), which was to consider the type inference process for an expression to produce two outputs: (1) an “expression artifact”, which is defined only by a prose description of its runtime semantics and therefore has no static type per se, and (2) a static type. This is actually a far better match for what actually happens in the CFE (expression artifacts are kernel Expressions, and although there is an `Expression.getStaticType` method in the kernel, Johnni has told me that it’s not always correct, and the CFE mostly disregards it in favor of its own independently-calculated static type). But it made the spec a lot more verbose, and I think it contributed to confusion about whether I was trying to make subsumption explicit in the language of elaborated expressions.

I’m going to leave this as is for now, but I’m working on a follow-up commit in which I add specifications for things like method invocations; I’ll experiment with your proposed invariant (any elaborated expression is either an expression of the source program, or it's a new form) in that PR.

Reply:
I'm OK with keeping `m_1.call` as the elaborated expression produced by type inference. I'd like to highlight the `.call` as being synthetic in some way, but if we can do that typographically it's probably fine.

However, "every elaborated expression has a well-defined static type" is the end result of the type inference algorithm, which we are in the middle of.
It's (I think) a recursive/inductive algorithm on expressions, where type is inferred on subexpressions first, resulting in elaborated expressions with static types (for the elaborated expression and every elaborated subexpression), then those elaborated expressions and static types are used to infer the elaborated expression and static type of the parent expression, maybe filling in missing type arguments and wrapping it in coercions. Maybe even some desugaring using `@...`, although I'd *prefer* to avoid it.

So "when the elaborated expression uses standard Dart syntax" does not apply to `m_1.call`, because `m_1` is an elaborated expression which may not use standard Dart syntax.
The `m_1.call` is not the result of a recursive call to type inference, so it hasn't been assigned a static type by that, it's a completely new elaborated expression which we have willed into existence from nothing, and it has no inherent static type.

We *definitely* don't want to do type inference on `m_1.call`, doing inference on `m_1` recursively (now in receiver position, if it's even a valid unelaborated expression any more).

That is: a recursive invocation of type inference returns an elaborated expression where every subexpression has been assigned a static type. The current invocation needs to do the do the same, so when it returns an alaborated expression, that, and all subexpressions, must also have been assigned static types. And someone needs to do that assigning - a static type is not an inherent property of an expression, it's the explicit result of running inference on the expression, and here we are implementing the inference which assigns the static type.

The right thing to do is to say that:
* We introduce an *new* elaborated expression `m_1.call`. This follows the grammatical rules for elaborated expressions (which is the grammar of expressions where `<primary>` has been extended with the new syntactic forms).
* We immediately assign this `m_1.call` the static type `U` (the function type of the `call` method of `T_1`, the static type of `m_1`, which inductively has a static type since it's an existing elaborated expression).
* Then `m_2` is `m_1`.

That is, when we introduce a new elaborated expression, which is not the result of running type inference on an unelaborated expression, then we also have to assign its static type. (Then we can still argue that that static type is sound.)

Reply:
> I'm OK with keeping `m_1.call` as the elaborated expression produced by type inference. I'd like to highlight the `.call` as being synthetic in some way, but if we can do that typographically it's probably fine.

I'd like to push back on this. What would we be trying to communicate by highlighting that `.call` is synthetic? The fact that `.call` doesn't appear in the user's source code in this case is obvious. Are we trying to highlight that coercion has inserted an operation? Again, I think that's obvious.

> However, "every elaborated expression has a well-defined static type" is the end result of the type inference algorithm, which we are in the middle of. It's (I think) a recursive/inductive algorithm on expressions, where type is inferred on subexpressions first, resulting in elaborated expressions with static types (for the elaborated expression and every elaborated subexpression), then those elaborated expressions and static types are used to infer the elaborated expression and static type of the parent expression, maybe filling in missing type arguments and wrapping it in coercions. Maybe even some desugaring using `@...`, although I'd _prefer_ to avoid it.

Ok, you have a good point. I think I was trying to avoid scope creep in this document by letting the rules for static typing be largely defined elsewhere, but it's really not panning out that way, is it? Not only do I have to define static type for the new constructs I'm introducing, but I'm also breaking my own principle and introducing pointless differences between the spec and the implementations (because the implementations _do_ compute static types as a by-product of the type inference process).

I've re-worked the PR so that every rule for expression inference explicitly specifies the static type of the resulting elaborated expression, and I've clarified that the rules in this document for determining the static type of an expression are meant to supersede the rules currently in `dartLangSpec.tex`.